test/ExecutionFramework.cpp

```
@@ -50,25 +55,294 @@ string getIPCSocketPath()
+		if (Account* acc = account(_addr))
+			return acc->storage[_key];
+		return {};
+	}
+
+	evmc_storage_status set_storage(
+		evmc_address const& _addr,
+		evmc_bytes32 const& _key,
+		evmc_bytes32 const& _value
+	) noexcept final
+	{
+		evmc_bytes32 previousValue = m_state.accounts[_addr].storage[_key];
+		m_state.accounts[_addr].storage[_key] = _value;
+
+		// TODO there are more possible values, especially "MODIFIED_AGAIN".
+		return previousValue == _value ? EVMC_STORAGE_UNCHANGED : EVMC_STORAGE_MODIFIED;
+	}
+
+	evmc_uint256be get_balance(evmc_address const& _addr) noexcept final
+	{
+		if (Account const* acc = account(_addr))
+			return acc->balance;
+		return {};
+	}
+
+	size_t get_code_size(evmc_address const& _addr) noexcept final
+	{
+		if (Account const* acc = account(_addr))
+			return acc->code.size();
+		return 0;
+	}
+
+	evmc_bytes32 get_code_hash(evmc_address const& _addr) noexcept final
+	{
+		if (Account const* acc = account(_addr))
+			return acc->codeHash;
+		return {};
+	}
+
+	size_t copy_code(
+		evmc_address const& _addr,
+		size_t _codeOffset,
+		uint8_t* _bufferData,
+		size_t _bufferSize
+	) noexcept final
+	{
+		// TODO is this supposed to fill with zero bytes?
+		Account const* acc = account(_addr);
+		if (!acc)
+			return 0;
+		size_t i = 0;
+		for (; i < _bufferSize && _codeOffset + i < acc->code.size(); i++)
+			_bufferData[i] = acc->code[_codeOffset + i];
+		return i;
+	}
+
+	void selfdestruct(evmc_address const& _addr, evmc_address const& _beneficiary) noexcept final
+	{
+		m_state.accounts[_beneficiary].balance = m_state.accounts[_addr].balance;
+		m_state.accounts.erase(_addr);
+	}
+
+	evmc::result call(evmc_message const& _message) noexcept final
+	{
+		u256 value{convertFromEVMC(_message.value)};
+		Account* sender = account(_message.sender);
+		// TODO are we responsible for checking balance?
+		if (!sender)// || u256(convertFromEVMC(sender->balance)) < value)
+		{
+			evmc_result res{};
+			// TODO correct to use revert here?
```

This should not happen, rather report an internal error.